homeassistant/components/ozw/__init__.py

```diff
@@ -69,16 +75,61 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     data_nodes = {}
     data_values = {}
     removed_nodes = []
+    manager_options = {"topic_prefix": f"{TOPIC_OPENZWAVE}/"}
 
-    @callback
-    def send_message(topic, payload):
-        mqtt.async_publish(hass, topic, json.dumps(payload))
+    if entry.unique_id is None:
+        hass.config_entries.async_update_entry(entry, unique_id=DOMAIN)
+
+    if entry.data.get(CONF_USE_ADDON):
+        # Do not use MQTT integration. Use own MQTT client.
+        host = entry.data.get(CONF_HOST)
+
+        if not host:
+            # Retrieve discovery info from the OpenZWave add-on.
+            discovery_info = (
+                await hass.components.hassio.async_get_addon_discovery_info(
+                    "core_zwave"
+                )
+            )
+
+            if not discovery_info:
+                _LOGGER.error("Failed to get add-on discovery info")
+                return False
+
+            discovery_info_config = discovery_info["config"]
+            addon_discovery_info = {
+                CONF_HOST: discovery_info_config["host"],
+                CONF_PORT: discovery_info_config["port"],
+                CONF_USERNAME: discovery_info_config["username"],
+                CONF_PASSWORD: discovery_info_config["password"],
+            }
+            hass.config_entries.async_update_entry(
```

Comment:
I'm a bit confused about why this part is needed? The `async_step_hassio` in the config flow updates the entry already when the SU pushes configuration, right?

Reply:
The discovery process in the hassio integration, that starts the config flow, runs on home assistant start event. So existing config entries will be loaded before that, and we need to get the discovery info to complete the entry setup.

Also, we currently do not get the discovery info in the user started flow. So the discovery flow will not update a new config entry that is created from the user step, since the discovery flow will be closed when creating the entry from the user flow. We could get the discovery info manually in the user flow, but since we need to do it here in the entry setup anyway, I've just added it here.

Reply:
I agree that it would be nice to condense this into one place, preferably having the config entry data ready before entry setup, but our current procedures limits us. Something to think about for the future, I guess.

Reply:
> The discovery processing of add-on info in the hassio integration, that starts the config flow, runs on home assistant start event. So existing config entries will be loaded before that, and we need to get the discovery info to complete the entry setup.

That should not be an issue. If the entry fails, the ConfigEntryNotReady is thrown. Meanwhile, the entry gets updated by the discovery process and on the next entry setup, it will succeed. Note; that the discovery data generally doesn't change; so this is all quite rare to happen.

> Also, we currently do not get the discovery info in the user started flow.

Hmmm right, that is a good point. However, we can get into other running flows and their context. (snatching its data?). Not the prettiest in the world probably either...

Reply:
So we should pick a way. 

- Always get credentials inside `async_setup_entry` and use them right away
- Store credentials in config entry and have `hassio_discovery` update it, raise ConfigEntryNotReady if credentials not stored yet on first launch

Reply:
The latter approach is cleaner code wise, but will raise an error for the case where the old config entry isn't updated yet. I like having all the data before the entry setup, so I prefer this.

Downside is that we store credentials in the config entry data.

Reply:
Those credentials, in this case, only work on an internal connection (within the hassio docker network). I prefer the latter approach in Paulus his list; mainly because it rarely changes.

Reply:
Good, we agree! I'll use that approach then.

Reply:
Now I'm thinking it's easier to do the other approach to solve the case with new user flows that closes the discovery flow.

Reply:
I settled for the other approach (1) as that was easier.